libraries/AC_AttitudeControl/AC_WeatherVane.cpp

```diff
@@ -0,0 +1,216 @@
+    switch (get_direction()) {
+        case Direction::OFF:
+            return 0.0;
+
+        case Direction::NOSE_OR_TAIL_IN:
+            output = roll_cdeg * _gain;
+
+            if (pitch_cdeg > 0) {
+                output *= -1.0;
+            }
+            break;
+
+        case Direction::NOSE_IN:
+            output = (fabsf(roll_cdeg) + MAX(pitch_cdeg,0.0)) * _gain;
+
+            // Yaw in the direction of the lowest 'wing'
+            if (roll_cdeg < 0) {
+                output *= -1.0;
+            }
+            break;
+
+        case Direction::SIDE_IN:
+            output = pitch_cdeg * _gain;
+
+            if (roll_cdeg < 0) {
+                output *= -1.0;
+            }
+            break;
+
+    }
+
+    // Don't activate weather vane if less than deadzone angle
+    float angle = get_direction() == Direction::SIDE_IN ? pitch_cdeg : roll_cdeg;
+    if (fabsf(angle) < _min_dz_ang_deg*100.0 && !(pitch_cdeg > _min_dz_ang_deg*100.0 && get_direction() == Direction::NOSE_IN)) {
+        output = 0.0;
+    }
+
+    // Force the controller to relax.  This can be called when landing.
+    if (should_relax) {
+        output = 0.0;
+        // Always reset should relax. Maintain relaxed condition by persistant calls to relax()
+        should_relax = false;
+    }
+
+    // Slew output
+    last_output = 0.98 * last_output + 0.02 * output;
+
+    return last_output;
+}
+
+// Called on an interrupt to reset the weathervane controller
+void AC_WeatherVane::reset(void)
+{
+    last_output = 0;
+    active_msg_sent = false;
+    should_relax = false;
+    first_activate_ms = 0;
+}
+
+// Returns true if the vehicle is in a condition whereby weathervaning is allowed
+// pilot_yaw can be an angle or a rate or rcin from yaw channel. It just needs to represent a pilot's request to yaw the vehicle.
+bool AC_WeatherVane::should_weathervane(const int16_t roll_cdeg, const int16_t pitch_cdeg, const int16_t pilot_yaw, const float hgt)
+{
+    // Check enabled
+    if (get_direction() == Direction::OFF){
+        reset();
+        return false;
+    }
+
+    const uint32_t now = AP_HAL::millis();
+
+    // Don't fight pilot inputs
+    if (pilot_yaw != 0) {
+        last_pilot_input_ms = now;
+        reset();
+        return false;
+    }
+
+    // Only allow weather vaning if no input from pilot in last 2 seconds
+    const uint32_t buffer_time_ms = 2000;
+    if (now - last_pilot_input_ms < buffer_time_ms) {
+        reset();
+        return false;
+    }
+
+    // Check if we are above the minimum height to weather vane
+    if (_min_height.get() > 0 && hgt <= _min_height.get()) {
+        reset();
+        return false;
+    }
+
+    // Check if we meet the horizontal velocity thresholds to allow weathervaning
+    const AP_AHRS &ahrs = AP::ahrs();
+    Vector3f vel_ned;
+    if (!ahrs.get_velocity_NED(vel_ned) || ((_max_vel_xy.get() > 0) && (sqrtf(vel_ned.x*vel_ned.x + vel_ned.y*vel_ned.y) > _max_vel_xy.get()))) {
```

Comment:
if your only using it one you can also do this
```suggestion
    if (!AP::ahrs().get_velocity_NED(vel_ned) || ((_max_vel_xy.get() > 0) && (sqrtf(vel_ned.x*vel_ned.x + vel_ned.y*vel_ned.y) > _max_vel_xy.get()))) {
```
and skip the extra refence.

Reply:
Done